src/axom/primal/operators/detail/clip_impl.hpp

```diff
@@ -671,6 +671,131 @@ AXOM_HOST_DEVICE Polyhedron<T, NDIMS> clipTetrahedron(
+ *          will have a non-positive and/or unexpected area.
+ *
+ * \warning If tryFixOrientation flag is false and some of the shapes have
+ *          a negative signed area, the returned Polygon
+ *          will have a non-positive and/or unexpected area.
+ */
+AXOM_SUPPRESS_HD_WARN
+template <typename T, axom::primal::PolygonArray ARRAY_TYPE, int MAX_VERTS>
+AXOM_HOST_DEVICE Polygon<T, 2, ARRAY_TYPE, MAX_VERTS> clipPolygonPolygon(
+  const Polygon<T, 2, ARRAY_TYPE, MAX_VERTS>& subjectPolygon,
+  const Polygon<T, 2, ARRAY_TYPE, MAX_VERTS>& clipPolygon,
+  double eps = 1.e-10,
+  bool tryFixOrientation = false)
+{
+  using PlaneType = Plane<T, 2>;
+  using PointType = Point<T, 2>;
+  using SegmentType = Segment<T, 2>;
+  using PolygonType = Polygon<T, 2, ARRAY_TYPE, MAX_VERTS>;
+
+  PolygonType outputList = subjectPolygon;
+  PolygonType planePoints = clipPolygon;
+
+  if(tryFixOrientation)
+  {
+    PolygonType tempPolygon;
+    if(subjectPolygon.signedArea() < 0)
+    {
+      for(int i = 0; i < subjectPolygon.numVertices(); i++)
+      {
+        tempPolygon.addVertex(
+          subjectPolygon[subjectPolygon.numVertices() - i - 1]);
+      }
+      outputList = tempPolygon;
+      tempPolygon.clear();
+    }
+
+    if(clipPolygon.signedArea() < 0)
+    {
+      for(int i = 0; i < clipPolygon.numVertices(); i++)
+      {
+        tempPolygon.addVertex(clipPolygon[clipPolygon.numVertices() - i - 1]);
+      }
+      planePoints = tempPolygon;
+      tempPolygon.clear();
+    }
+  }
+
+  int numClipEdges = planePoints.numVertices();
+
+  // Iterate through edges of clip polygon, represented as planes
+  for(int i = 0; i < numClipEdges; i++)
+  {
+    PlaneType plane =
+      make_plane(planePoints[i], planePoints[(i + 1) % numClipEdges]);
+
+    PolygonType inputList = outputList;
+    outputList.clear();
+
+    for(int i = 0; i < inputList.numVertices(); i++)
+    {
+      PointType current_point = inputList[i];
+      PointType prev_point =
+        inputList[(i - 1) == -1 ? (inputList.numVertices() - 1) : (i - 1)];
+
+      T seg_param;
+      PointType intersecting_point;
+      SegmentType subject_edge(prev_point, current_point);
+
+      if(intersect(plane, subject_edge, seg_param))
+      {
+        intersecting_point = subject_edge.at(seg_param);
+      }
+
+      int cur_p_orientation = plane.getOrientation(current_point, eps);
+      int prev_p_orientation = plane.getOrientation(prev_point, eps);
+
+      if(cur_p_orientation == ON_POSITIVE_SIDE)
+      {
+        // Handles the edge case of 3 consecutive vertices with orientations
+        // ON_POSITIVE_SIDE, ON_BOUNDARY, ON_POSITIVE. Default algorithm
+        // check (prev_p_orientation != ON_POSITIVE_SIDE) results in the
+        // vertex on the boundary being added twice.
+        if(prev_p_orientation == ON_NEGATIVE_SIDE ||
+           (prev_p_orientation == ON_BOUNDARY &&
+            intersecting_point != outputList[outputList.numVertices() - 1]))
```

Comment:
Are there ever cases where this will be called and the curves did not intersect (e.g. intersecting point was default initialized)?

It seems like we might want to move the intersect call into this check as follows.
(I think the following change should be equivalent and more efficient, but please check):
```cpp

      const int cur_p_orientation = plane.getOrientation(current_point, eps);
      const int prev_p_orientation = plane.getOrientation(prev_point, eps);

      if(cur_p_orientation == ON_POSITIVE_SIDE)
      {
        // Handles the edge case of 3 consecutive vertices with orientations
        // ON_POSITIVE_SIDE, ON_BOUNDARY, ON_POSITIVE. Default algorithm
        // check (prev_p_orientation != ON_POSITIVE_SIDE) results in the
        // vertex on the boundary being added twice.
        if(prev_p_orientation == ON_NEGATIVE_SIDE ||
           (prev_p_orientation == ON_BOUNDARY)
        { 
          T seg_param;
          SegmentType subject_edge(prev_point, current_point);

           if(intersect(plane, subject_edge, seg_param))
           {
              const auto intersecting_point = subject_edge.at(seg_param);
              if(intersecting_point != outputList[outputList.numVertices() - 1]))
              {
                 outputList.addVertex(intersecting_point);
              }
           }
         }
         outputList.addVertex(current_point);
      }
      else if(prev_p_orientation == ON_POSITIVE_SIDE)
      {
        outputList.addVertex(intersecting_point);
      }
```

Reply:
Also, do we have tests that cover all these cases?

Reply:
> Are there ever cases where this will be called and the curves did not intersect (e.g. intersecting point was default initialized)?

No, there will always be an intersection when the if condition is satisfied at this point.
The main issue was the edge case I mentioned in the comments, which can result in a point being added twice in that specific scenario.

> It seems like we might want to move the intersect call into this check as follows.
(I think the following change should be equivalent and more efficient, but please check):

I gave it a try, but there is a compilation failure as `intersection_point` is not defined at the `else-if` block:
```
      else if(prev_p_orientation == ON_POSITIVE_SIDE)
      {
        outputList.addVertex(intersecting_point); // intersecting_point not defined
      }
```

I left the branching structure as-is, since the intersection point needs to be known by both branch paths.
The Sutherland–Hodgman algorithm should guarantee there is an intersection point in the `else-if` case, as when teh condition is satisfied the current point is either on the negative side or on the edge, while the previous point is on the positive side.

> Also, do we have tests that cover all these cases?

Yes, all branch paths are covered, including a test for the specific edge case mentioned.